File: python/middle/subsets_ii.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ret = [[]]
        if len(nums) == 0:
            return ret
        idx = 0
        tmp = []
        nums.sort()
        
        def dfs(index):
            if index >= len(nums):
                return
            for i in range(index, len(nums)):
                tmp.append(nums[i])
                if tmp not in ret:
                    test = tmp[:]
                    ret.append(test)
                dfs(i+1)
                tmp.pop()
                while( (i < len(nums)-1) and (nums[i] == nums[i+1])):
                    i += 1
            return
        dfs(idx)
        return ret

File: cpp/middle/subsets_ii.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> ret;
        vector<int> tmp;
        ret.push_back(tmp);
        if(nums.size() == 0)
            return ret;
        sort(nums.begin(), nums.end()); // inorder to skip the duplicate answer, we need to sort the input array
        int idx = 0;
        dfs(ret, tmp, nums, idx);
        return ret;
    }
    
    void dfs(vector<vector<int>>& ret, vector<int>& tmp, vector<int>& nums, int idx)
    {
        if(idx >= nums.size())
            return;
        for(int i = idx; i < nums.size(); ++i)
        {
            tmp.push_back(nums[i]);
            ret.push_back(tmp);
            dfs(ret, tmp, nums, i+1);
            tmp.pop_back();
            while(i < nums.size()-1 && nums[i] == nums[i+1])
                ++i;
        }
    }
};

Conclusion:

Middle level algorithm issue, the time complexity is the same with subsets, which is: O(n*2^n), space complexity is: O(n)

